### PR TITLE
Feature digital io

### PR DESCRIPTION
**Pull request by Lurkerpas**

### CODE REVIEW

File: debugger/io.c [hunk 1]
```diff
@@ -19,6 +19,20 @@
...
+static void digitalIoToString(char* const output, const uint8_t mask, const uint8_t x)
+{
+    for (size_t i = 0; i < BIT_COUNT; i++)
+    {
+        const uint8_t bit = 1 << i;
+        const bool valid = (mask & bit) != 0;
+        const bool one = (x & bit) != 0;
+        output[i] = valid ? (one ? '1' : '0') : 'X';
+    }
+    output[BIT_COUNT + 1] = 0;
```
Comment (hcorg) [inline debugger/io.c]: `'\0'` (IMHO it clearly shows were are the "string manipulation" things)

File: debugger/io.c [hunk 2]
```diff
@@ -19,6 +19,20 @@
 #include "io.h"
 #include "websockets/packet_queue.h"
 
+#define BIT_COUNT 8
+
+static void digitalIoToString(char* const output, const uint8_t mask, const uint8_t x)
```
Comment (hcorg) [inline debugger/io.c]: one of rare occasions to use  known output size: `char output[BIT_COUNT + 1] `, should allow compiler to detect some bugs (hopefully)

File: debugger/io.c (hunk 1 above)
Comment (hcorg) [reply debugger/io.c]: also - why + 1?

File: main.cpp [hunk 3]
```diff
@@ -211,6 +215,31 @@ static bool closeUsciPipes(Emulator* const emu)
...
+    }
+    return true;
+}
+
+static bool closeDioPipes(Emulator* const emu)
+{
+    if (emu->port1_output_pipe != NULL)
+    {
+        fclose(emu->port1_output_pipe);
+        emu->port1_output_pipe = NULL;
```
Comment (hcorg) [inline main.cpp:240]: `_fd` = ???

mkay, I see it's a common pattern in this file to leave fd of closed file untouched.

**Review by hcorg — APPROVED** [review]

File: debugger/io.c (hunk 1 above)
Comment (rphh) [reply debugger/io.c]: null terminated string :)

File: debugger/io.c [hunk 4]
```diff
@@ -34,6 +48,27 @@ void print_serial (Emulator *emu, char *buf)
     }
 }
 
+void put_port1(Emulator* const emu, const uint8_t mask, const uint8_t x)
+{
+
```
Comment (rphh) [inline debugger/io.c]: style: empty line

**Review by rphh — APPROVED** [review]

File: debugger/io.c (hunk 1 above)
Comment (Lurkerpas) [reply debugger/io.c]: '\0' - Ack
+1 - a classical off-by-one, introduced when I switched to defines... Fixing

File: debugger/io.c (hunk 2 above)
Comment (Lurkerpas) [reply debugger/io.c]: Ack

File: debugger/io.c (hunk 4 above)
Comment (Lurkerpas) [reply debugger/io.c]: Ack

File: main.cpp (hunk 3 above)
Comment (Lurkerpas) [reply main.cpp:240]: fd is derived from the opened file, which is now getting closed;the condition checked in other places to use the pipe/pipe_fd is pipe != NULL

Comment (Lurkerpas) [reply main.cpp:240]: But ok, added cleaning up fd, just in case

**Pull request merged**